Update configuration builder settings classes setup, fallback to conf…

MishaDemianenko opened the pull request:
…ig map in case if settings classes is missing.

MishaDemianenko committed:
Update configuration builder settings classes setup, fallback to config map in case if settings classes is missing.

MishaDemianenko commented:
Hi @zhenlineo can you please take a look, 
- Update `ConfiguratorWrappingConfigurationBuilder`to use settings classes 
- Fall back to config map in  case if `ConfigWrappingConfiguration` can't find setting for some reason.

MishaDemianenko commented:
#4851

jakewins commented on community/server/src/main/java/org/neo4j/server/configuration/ConfigurationBuilder.java:
```diff
@@ -137,4 +141,13 @@ public Configuration configuration()
 
     }
+
+    class DefaultSettingsClassesHolder
```
How come this static method is wrapped inside this inner class?

MishaDemianenko replied:
Not really sure that i fully understand the question :), The idea of that class to provide common place where different configurations can get settings classes

zhenlineo replied:
I also feels a bit strange to keep the static method in a class here. Though I have no better idea where it should go, maybe Jake has some good idea? @jakewins

zhenlineo replied:
https://github.com/neo4j/neo4j/blob/2.3/community/server/src/main/java/org/neo4j/server/configuration/ServerConfigFactory.java#L117

jakewins replied:
Why not just put the static method in the ConfigurationBuilder directly? It's already in that class, just wrapped in this inner class.

MishaDemianenko replied:
Because we can't have static methods in interfaces prior java 8 :)

zhenlineo commented on community/server/src/main/java/org/neo4j/server/configuration/ConfigWrappingConfiguration.java:
```diff
@@ -60,31 +60,15 @@ public boolean containsKey( String key )
-        }
-        return notNullKeys.iterator();
+        Set<String> propertyKeys = new HashSet<>(config.getParams().keySet());
+        propertyKeys.addAll( getRegisteredSettings().keySet() );
```
I feel this changes its previous behaviour. Before, as indicated by the comments, this method will only keep the keys of `properties which have been assigned default values` and the key of properties defined by users.

MishaDemianenko commented:
@zhenlineo restore check for empty properties

zhenlineo commented on community/server/src/main/java/org/neo4j/server/configuration/ConfigWrappingConfiguration.java:
```diff
@@ -60,31 +60,21 @@ public boolean containsKey( String key )
-        // only keep the properties which have been assigned default values or user-defined values
-        Set<String> notNullKeys = new HashSet<>();
-        for ( String key : staticKeys )
+        Set<String> propertyKeys = new HashSet<>(config.getParams().keySet());
```
How about also keep the comments in case we forgot to filter properties without a default value in the future? Besides, plz use the default code style, you are missing spaces here :)

zhenlineo commented on community/server/src/test/java/org/neo4j/server/configuration/ConfigWrappingConfigurationTest.java:
```diff
@@ -67,13 +69,30 @@ public void shouldGetPropertyInRightFormat() throws Exception
+    @Test
+    public void shouldAbleToAccessRegisteredPropertyByName()
+    {
+        Config config = new Config(new HashMap<String, String>(  ), getDefaultSettingsClasses());
```
FYI, `(  )` should change to `()`

zhenlineo commented on community/server/src/test/resources/META-INF/services/org.neo4j.server.plugins.PluginLifecycle:
```diff
@@ -0,0 +1 @@
+org.neo4j.server.modules.ExtensionInitializerTest$PropertyCollectorPlugin
```
I do not quite understand why we have this file?

MishaDemianenko replied:
That is for `ExtensionInitializerTest`. `ExtensionInitializer` uses `Service.load( PluginLifecycle.class )` to load plugins and that is a way how those will be found.

MishaDemianenko commented:
@zhenlineo comment and formatting

zhenlineo commented:
It looks good to me. Let's see if Jake has something else.

chrisvest commented:
@MishaDemianenko Not sure if the test failures are relevant.

MishaDemianenko commented:
@chrisvest  sure that they are not, but can trigger yet another build to be incredibly confident :)

MishaDemianenko commented:
@zhenlineo ping

zhenlineo commented:
Need a rebase

MishaDemianenko commented:
@zhenlineo rebased and updated.
@jakewins saw your new `ServerConfigFactory` and remove mine new static holder class from interface, updated `ServerConfigFactory` method to provide config classes.
